Python/ceval.c

```diff
@@ -2493,6 +2496,18 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             return retval;
         }
 
+        TARGET(RETURN_VALUE) {
+            PyObject *retval = TOP();
+            frame->f_state = FRAME_RETURNED;
+            /* FIXME(lpereira): The stack pointer and f_state will be set to the same
+             * value here, and when jumping to return_value_without_tracing above. This
+             * is redundant, but saves us from copying some code. */
```

Comment:
Honestly, I would just duplicate the code. You also redundantly have an extra variable `retval`, which is only used in the `TRACE_FUNCTION_EXIT()` macro. Maybe something needs to be refactored again here.

Reply:
I did some refactoring and the variable that's only used when tracing is enabled is now gone.  (It's now inside the macro, which now takes a parameter.)

I dislike both copying code and doing what has been done in this PR, to be honest... while I do have a slight preference to the way I've done things as it's quite a bit of code that don't need to be copied around.

Reply:
Another possible option: split the opcode in two? So we have one opcode which does the tracing, and one which does the return.

The original code would be `EXIT_FUNCTION + RETURN_VALUE`, and the quickening step would replace `EXIT_FUNCTION` with `NOP` (or, even better, `RETURN_VALUE`).

I'm *pretty* sure that could work correctly here, but I might be missing some edge case where an eval break or generator `throw()` could leave us in a weird state where we trace a return that never happens (or something).

Reply:
Hm, never mind. I don't know how much that would actually clean this up.